text/0000-dyn-assoc.md

```diff
@@ -0,0 +1,371 @@
+
+This issue also exists when using the generics on types:
+```rust
+trait Foo {
+    const MY_NAME: &'static str;
+}
+struct MyBox<T: Foo + ?Sized>(Box<T>);
+
+impl<T: Foo + ?Sized> MyBox<T> {
+    fn get_name() -> &'static str {
+        T::MY_NAME
+    }
+}
+```
+In this case, the compiler **detects that the type is using the associated members of `T`** and, if we then try to give `dyn Foo` to it, produces the a similar error: "cannot use dyn Trait as a generic parameter because the type requires a trait object to (call associated function|use associated constant)", with a similar highlight as above.
+
+If we have a trait object, however, and we want to use compile-time generics, we can use it directly as the generic parameter:
+```rust
+trait Foo {
+    const CONST: usize;
+    
+    fn assoc();
+}
+// Long story short, we're taking a generic argument
+// without a vtable, but we need to do dynamic dispatch.
+fn foo<T: Foo + ?Sized>() {
+    let _ = T::CONST;
+    T::assoc();
+}
+
+struct DefaultFoo;
+impl Foo for DefaultFoo {
+    const CONST: usize = 42;
+    
+    fn assoc() {
+        println!("I'm doing great, thank you!");
+    }
+}
+
+fn main() {
+    // We'd have a constructor here if the implementor wasn't a unit type:
+    let tobject: Box<dyn Foo> = Box::from(DefaultFoo);
+    foo::<tobject.as_ref()>();
+}
+```
+Here, we're saying to the compiler: "take this trait object and use it to resolve all usage of associated constants/functions in this function I'm calling". Same with types with generic arguments.
+
+### The vtable ambiguity rule
+Consider the following code:
+```rust
+trait Foo {
+    const CONST: usize;
+}
+fn foo<T: Foo + ?Sized>(t1: &T, t2: &T) -> usize {
+    T::CONST
+}
+struct DefaultFoo;
+impl Foo for DefaultFoo {
+    const CONST: usize = 42;
+}
+struct OtherFoo;
+impl Foo for OtherFoo {
+    const CONST: usize = 87;
+}
+
+fn main() {
+    let tobject1: Box<dyn Foo> = Box::from(DefaultFoo);
+    let tobject2: Box<dyn Foo> = Box::from(OtherFoo);
+
+    foo(tobject1.as_ref(), tobject2.as_ref());
+}
+```
+What happens here? More specifically, what does `T::CONST` do in `foo`, where does it resolve to? We can't implicitly infer the vtable used, since there are two instead of one. The only way to solve this is to introduce the vtable ambiguity rule: "a vtable should either be specified for the generic parameter or should be trivial to infer". Not following the rule should produce an error.
+
+There are two ways to resolve the issue:
+- Introduce an equality bound
+- Explicitly use one of the vtables
+
+Let's look at both options in order. An eqality bound would look like this:
+```rust
+trait Foo {
+    const CONST: usize;
+}
+fn foo<T: Foo + ?Sized>(t1: &T, t2: &T) -> usize
+where t1::impl == t2::impl {
+    T::CONST
+}
```

Comment:
This bound would be impossible to satisfy in real code. We can't know if the underlying types are the same, as this is the entire point of type erasure. Not even at runtime, otherwise you would have unsoundess due to lifetimes.

```rust
impl Foo for &i32 {}
```

At runtime you have no way of knowing what choice of lifetime is used on `&i32`, so even if you check that the v-tables are the same that doesn't guarantee that the types are indeed the same.